android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java

```diff
@@ -264,38 +279,59 @@ public void getAccounts(Callbacks.Callback1<AccountInfo[]> callback1) {
 
     public void addAccount(String accountName, @CoinType.EnumType int coinType,
             Callbacks.Callback1<Boolean> callback) {
-        final AccountInfo[] finalAccountInfos =
+        final AccountInfo[] finalAccountInfo =
                 WalletUtils.getAccountInfosFromKeyrings(_mKeyringInfosLiveData.getValue())
                         .toArray(new AccountInfo[0]);
-        mKeyringService.addAccount(accountName, coinType, result -> {
-            if (result) {
-                boolean hasExistingAccountType = false;
-                for (AccountInfo accountInfo : finalAccountInfos) {
-                    hasExistingAccountType = (accountInfo.coin == coinType);
-                    if (hasExistingAccountType) break;
-                }
-                if (!hasExistingAccountType) {
-                    mKeyringService.getKeyringInfo(
-                            getSelectedCoinKeyringId(coinType), updatedKeyringInfo -> {
-                                for (AccountInfo accountInfo : updatedKeyringInfo.accountInfos) {
-                                    if (accountInfo.coin == coinType) {
-                                        setSelectedAccount(accountInfo.address, coinType);
-                                        break;
-                                    }
-                                }
-                            });
-                }
-            }
-            mCryptoSharedActions.updateCoinType();
-            mCryptoSharedActions.onNewAccountAdded();
-            callback.call(result);
-        });
+        if (coinType == CoinType.FIL) {
+            mKeyringService.addFilecoinAccount(
+                    accountName, BraveWalletConstants.FILECOIN_MAINNET, result -> {
+                        handleAccountResult(result, finalAccountInfo, coinType, callback);
+                    });
+        } else {
+            mKeyringService.addAccount(accountName, coinType, result -> {
+                handleAccountResult(result, finalAccountInfo, coinType, callback);
+            });
+        }
     }
 
     public void isWalletLocked(Callbacks.Callback1<Boolean> callback) {
         mKeyringService.isLocked(isWalletLocked -> callback.call(isWalletLocked));
     }
 
+    private void handleAccountResult(boolean result, AccountInfo[] finalAccountInfo,
```

Comment:
nit: I think we should avoid multiple methods like `handle...` because it makes the flow/code obscure. Preferably use `handleAddAccountResult`.
Same for `handleSelectedAccount.

Reply:
There are two methods that handle the result of the callbacks: `handleSelectedAccount` and `handleAccountResult`. 

If you find the code unclear I'm happy to drop the `handle` prefix as you suggested. Do you have a preference for the new method names?

What about `addAccountResult` and `getSelectedAccountResult`?
Another option could be `addAccountResponse` and `getSelectedAccountResponse`.
Open to suggestions here.

Reply:
> handleAddAccountResult though would be better to use updateSelectedAccountAndState I think.

Perfect then, I'll rename it that way.

> Btw, would be better to rename finalAccountInfo => finalAccountInfos for an array.

`accountInfoArray`? How does that sound? (Not a huge fun of `infos`, it feels wrong to me, sorry to nitpick around this).